### PR TITLE
[AIRFLOW-4490] dag_run.conf should be an empty dictionary by default instead of None

### PR DESCRIPTION
**Pull request by yampelo**
Changes DagRun.conf to be an empty dictionary by default, rather than None.

Make sure you have checked _all_ steps below.

### Jira

- [ ] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-4490\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-4490
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-4490\], code changes always need a Jira issue.
  - In case you are proposing a fundamental code change, you need to create an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)).
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Description

- [ ] Here are some details about my PR, including screenshots of any UI changes:

### Tests

- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [ ] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [x] Passes `flake8`

### CODE REVIEW

File: airflow/models/dagrun.py [hunk 1]
```diff
@@ -54,7 +54,7 @@ class DagRun(Base, LoggingMixin):
     _state = Column('state', String(50), default=State.RUNNING)
     run_id = Column(String(ID_LEN))
     external_trigger = Column(Boolean, default=True)
-    conf = Column(PickleType)
+    conf = Column(PickleType, default={})
```
Comment (potiuk) [inline airflow/models/dagrun.py:57]: I think it would be great to add a test case covering this empty dict behavior.

**Review by potiuk — CHANGES_REQUESTED** [review]

**Comment by yampelo** [comment]
Is there ever a case where the DagRun is not committed to that database immediately?  Defining the default value in the `Column` constructor (like this `Column(PickleType, default={})`) will only use the default value after an `INSERT` statement. That means that:
```python
dag_run = DagRun(....) # initialized without a conf 
dag_run.conf  == None # None at this point

session.add(dag_run)
session.commit() # Insert places default value

dag_run_in_db = Session.query(DagRun).first()
dag_run_in_db.conf == {} # conf is now the default value
```

**Comment by dstandish** [comment]
> Is there ever a case where the DagRun is not committed to that database immediately?

Intuitively, it seems likely impossible in actual execution.  But if you are experimenting in interactive console, as you have demonstrated, it is certainly possible. (And I do that sometimes when working out template behavior.)
Another alternative approach to this would be to leave default value alone, but just add a `get_conf` method on `DagRun`, just like there is now with `get_dag`.  `get_conf` could return an empty dict if there is no conf.
We could also add a `get_dag_run_conf` method to `TaskInstance`, which would also handle the case when there is no `dag_run`.  Then you could safely do something like 
`{{ ti.get_dag_run_conf.get('start_date') or ds }}` in templates.
Or maybe it would make more sense to define it so the method is conf param getter rather than conf getter: e.g. `{{ ti.dag_run_conf_get('start_date') or ds }}`.

**Comment by dstandish** [comment]
@yampelo another alternative:
If the goal is to just make templating work properly, maybe all we should do is change `get_template_context` so that `dag_run` gets a default of `{}` in the template context!
Here's the line: https://github.com/apache/airflow/blob/220c1621702a61b661b1a16c45944812330ca2cf/airflow/models/taskinstance.py#L1079

**Comment by dstandish** [comment]
@yampelo one more idea:
If the concern here is just templating behavior, we could just add a `dag_run_conf` attribute to the template context, which would be `{}` by default and the conf if one exists.  Then we could do, for example, `{{ dag_run_conf.get('start_date', ds) }}`.

**Comment by dstandish** [comment]
@yampelo ok last idea (probably) 
Maybe this is better... in `TaskInstance.get_template_context`, could initialize `dag_run` with `DagRun()` and set `dag_run.conf = {}`.
here: https://github.com/apache/airflow/blob/7dda0c120a336ac6c12e2de0ca850bbafdc71483/airflow/models/taskinstance.py#L1079

**Comment by potiuk** [comment]
Now seeing that commit is needed, that's not a good idea to get default. How about another idea @yampelo -> maybe you can add a new property (config?)  to DagRun which will return {} if configuration is None.

**Comment by ashb** [comment]
Closing this PR - feature may be needed but not like this. There are two alternative PRs open for this now.

**Pull request closed without merging**